lib/smart_name.rb

```diff
@@ -44,20 +60,31 @@ def banned_re
   #~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
   #~~~~~~~~~~~~~~~~~~~~~~ INSTANCE ~~~~~~~~~~~~~~~~~~~~~~~~~
 
-  attr_reader :simple, :parts, :key, :s
+  attr_reader :namespaced, :simple, :rootspace, :parts, :key, :s
   alias to_s s
 
   def initialize str
     @s = str.to_s.strip
-    @s = @s.encode('UTF-8') if RUBYENCODING
+    @s = @s.encode('UTF-8')
     @key = if @s.index(self.class.joint)
         @parts = @s.split(/\s*#{JOINT_RE}\s*/)
         @parts << '' if @s[-1,1] == self.class.joint
         @simple = false
-        @parts.map { |p| p.to_name.key } * self.class.joint
+        names = @parts.each_with_object([]) do |part, names|
+            partial_key= names.flatten.map{ |p| p.to_name.key } * self.class.joint
+            if names.length == 0 || SmartName[partial_key]
+              names << [part]
+            else
+              names[-1] << part
+            end
+          end
+        @namespaced = names.map{|ns| ns * self.class.joint}
+        @rootspace = namespaced.length < 2
+        parts.map { |p| p.to_name.key } * self.class.joint
       else
-        @parts = [str]
-        @simple = true
+        @namespaced = [s]
+        @parts = [s]
+        @simple = @rootspace = true
         str.empty? ? '' : simple_key
       end
```

Comment:
Yes, this branch is more of a spike to think about the issues.  Let's put any detailed discussion of what we need in related tickets.
I don't think we quite know what the API will look like, and even how the objects will be bound together.  I suspect we may have more than one name parser instance or something along those lines.  Each deck, language combo might have a parser instance.  Not sure yet.